tools/src/main/scala/scala/scalanative/build/package.scala

```diff
@@ -0,0 +1,208 @@
+
+import scala.sys.process.Process
+
+import build.IO.RichPath
+import nir.Global
+
+package object build {
+
+  type LinkerPath = linker.ClassPath
+  val LinkerPath = linker.ClassPath
+
+  type LinkerReporter = linker.Reporter
+  val LinkerReporter = linker.Reporter
+
+  type LinkerResult = linker.Result
+  val LinkerResult = linker.Result
+
+  type OptimizerDriver = optimizer.Driver
+  val OptimizerDriver = optimizer.Driver
+
+  type OptimizerReporter = optimizer.Reporter
+  val OptimizerReporter = optimizer.Reporter
+
+  /** Given the classpath and main entry point, link under closed-world
+   *  assumption.
+   */
+  def link(config: Config): LinkerResult = {
+    val result = config.logger.time("Linking") {
+      val chaDeps   = optimizer.analysis.ClassHierarchy.depends
+      val passes    = config.driver.passes
+      val passDeps  = passes.flatMap(_.depends).distinct
+      val deps      = (chaDeps ++ passDeps).distinct
+      val injects   = passes.flatMap(_.injects)
+      val mainClass = nir.Global.Top(config.entry)
+      val entry =
+        nir.Global
+          .Member(mainClass, "main_scala.scalanative.runtime.ObjectArray_unit")
+      val result =
+        (linker.Linker(config)).link(entry +: deps)
+
+      result.withDefns(result.defns ++ injects)
+    }
+
+    if (result.unresolved.nonEmpty) {
+      result.unresolved.map(_.show).sorted.foreach { signature =>
+        config.logger.error(s"cannot link: $signature")
+      }
+      throw new Exception("unable to link")
+    }
+    val classCount = result.defns.count {
+      case _: nir.Defn.Class | _: nir.Defn.Module | _: nir.Defn.Trait => true
+      case _                                                          => false
+    }
+    val methodCount = result.defns.count(_.isInstanceOf[nir.Defn.Define])
+    config.logger.info(
+      s"Discovered ${classCount} classes and ${methodCount} methods")
+
+    result
+  }
+
+  /** Link just the given entries, disregarding the extra ones that are
+   *  needed for the optimizer and/or codegen.
+   */
+  def linkRaw(config: Config, entries: Seq[nir.Global]): LinkerResult =
+    config.logger.time("Linking") {
+      linker.Linker(config).link(entries)
+    }
+
+  /** Transform high-level closed world to its lower-level counterpart. */
+  def optimize(config: Config,
+               assembly: Seq[nir.Defn],
+               dyns: Seq[String]): Seq[nir.Defn] =
+    config.logger.time(s"Optimizing (${config.driver.mode} mode)") {
+      optimizer.Optimizer(config, assembly, dyns)
+    }
+
+  /** Given low-level assembly, emit LLVM IR for it to the buildDirectory. */
+  def codegen(config: Config, assembly: Seq[nir.Defn]): Seq[Path] = {
+    config.logger.time("Generating intermediate code") {
+      scalanative.codegen.CodeGen(config, assembly)
+    }
+    val produced = IO.getAll(config.workdir, "glob:**.ll")
+    config.logger.info(s"Produced ${produced.length} files")
+    produced
+  }
+
+  def build(nativeLib: Path,
```

Comment:
Needs a big fat docstring. We expect this to be the main entry point so it should be very clear and very prominent that it's the one you want to use if you're implementing Scala Native support in your build tool. It would be nice to put this higher up in the file as well.